src/watchman_client.js

```diff
@@ -0,0 +1,492 @@
+// WatchmanWatcher.createOptions() method. 
+Object.defineProperty(WatchmanClient.prototype, 'wildmatch', {
+  get() { return this._wildmatch; }
+});
+
+/**
+ * Called from WatchmanWatcher (or WatchmanClient during reconnect) to create
+ * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+ * watchman.Client, to be handled here.
+ */
+WatchmanClient.prototype.subscribe = function(watchmanWatcher, root) {
+  let subscription;
+  let watcherInfo;
+  
+  return this._setupClient()
+    .then(() => {
+      watcherInfo = this._createWatcherInfo(watchmanWatcher);
+      subscription = watcherInfo.subscription;
+      return this._watch(subscription, root);
+    })
+    .then((resp) => this._clock(subscription))
+    .then((data) => this._subscribe(subscription))
+    .catch((error) => {
+      console.error("Failed creating subscription for root " + root);
+      throw error;
+    });
+};
+
+/**
+ * Remove the information about a specific WatchmanWatcher.
+ * Once done, if no watchers are left, clear the local vars,
+ * which will end the connection to the watchman.Client, too.
+ */
+WatchmanClient.prototype.closeWatcher = function(watchmanWatcher) {
+  let watcherInfos = Object.values(this._watcherMap);
+  let numWatchers = watcherInfos.length;
+
+  if (numWatchers > 0) {
+    let watcherInfo;
+    
+    for (let info of watcherInfos) {
+      if (info.watchmanWatcher === watchmanWatcher) {
+        watcherInfo = info;
+        break;
+      }
+    }
+
+    if (watcherInfo) {
+      delete this._watcherMap[watcherInfo.subscription];
+
+      numWatchers--;
+
+      if (numWatchers === 0) {
+        this._clearLocalVars();  // nobody watching, so shut the watchman.Client down.
+      }
+    }
+  }
+};
+
+/**
+ * Simple backoff-time iterator. next() returns times in ms.
+ * When it's at the last value, it stays there until reset()
+ * is called.
+ */
+WatchmanClient.prototype._backoffTimes = {
+  _times: [0, 1000, 5000, 10000, 60000],
+
+  _next: 0,
+
+  next() {
+    let val = this._times[_next];
+    if (this._next < this._times.length - 1) {
+      this._next++;
+    }
+    return val;
+  },
+
+  reset() {
+    this._next = 0;
+  }
+}
+
+/**
+ * Set up the connection to the watchman client. Return a promise
+ * that is fulfilled when we have a client that has finished the
+ * capabilityCheck. 
+ */
+WatchmanClient.prototype._setupClient = function() {
+  if (!this._clientPromise) {
+    this._clientPromise = new Promise((resolve, reject) => {
+      this._handleClientAndCheck(resolve, reject);
+    });
+  }
+
+  return this._clientPromise;
+};
+
+/**
+ * Handle the process of creating a client and doing a capability check and
+ * getting a valid response, then setting up local data based on that.
+ *
+ * This is split from _setupClient and _createClientAndCheck so it can
+ * provide the backoff handling needed during attempts to reconnect.
+ */
+WatchmanClient.prototype._handleClientAndCheck = function(resolve, reject) {
```

Comment:
as implemented this method poses some risks. If an error is thrown prior to invoking `resolve` or `reject` (loc:150 -> 160, or loc: 174) it will be dropped, and `clientPromise` may end up in a permanently "hung" state.